### ISSUE TITLE
Replay attacks by the leader and helper statefulness

### ISSUE THREAD

**Issue by csharrison** [abetterinternet/prio-documents#62]
**TL;DR both aggregators need to maintain state to enforce privacy guarantees. Seems like there is an explicit state/communication trade-off to allow this**

Filing this issue based on the design meeting today.

In the updated design in https://github.com/abetterinternet/prio-documents/pull/59 the leader is in a somewhat privileged position in that it receives reports/batches directly and is the primary entity that keeps track of state.

For differential privacy guarantees, we want to bound the number of times any record / batch can contribute to an output to the collector. This requires some statefulness that the leader could help facilitate. However, if the leader is colluding with the collector the following attack is possible:

1. Collector issues a query to the leader: "aggregate the last day of data"
2. The protocol proceeds as normal and the collector receives the aggregate data + noise
3. Collector issues a similar query to the (corrupt) leader, asking for that same data again
4. Leader ignores its state that would otherwise say not to aggregate this data again and communicates to the helper the same data as in (2)
5. The collector receives the same data with newly sampled noise, and can leverage averaging attacks to diminish privacy

This is all possible because keeping track of "budgets" is the sole responsibility of the leader in step (4). To mitigate this we would need both helper and leader to maintain this information independently, so that neither aggregator has an advantage or can reduce user privacy on their own.

To do this, the helper will be required to keep track of (some) state. Additionally, we can rely on the leader to keep some state on behalf of the helper in some way that maintains the integrity of the data and prevents mutation. I believe there is an explicit tradeoff here in terms of (a) statefulness of the helper and (b) communication cost with the leader to maintain the auxiliary state.

On one extreme is the maximum statefulness in which all state is maintained by the helper (i.e. the helper and leader keep similar privacy budget databases). On the other extreme is maximum statelessness which would entail something like the helper keeping track of a single counter which "versions" the helper's database stored on the leader. The entire database would be communicated to the helper during the protocol, and a new version would be pushed to the leader for storage.

How big would this database be? This is a bit of an open question but it will scale proportionally to the number of _partitions_ of the data we would support querying independently (without impacting budget of other partitions). In the design outlined in https://github.com/WICG/conversion-measurement-api/blob/main/AGGREGATE.md we would partition based on:
 - Some time window (e.g. hour)
 - The collector's identity
 - Some application specific key which is specified by the client

I could imagine this being large in practice though I don't have exact numbers. One natural compromise design between these two extreme options would be to partition the auxiliary databases stored on the leader per one of these keys (e.g. collector identity). This would entail the helper keeping c counters, one for each collector. Then the aux state transferred during each protocol is just the state for the relevant collector. You could presumably do this for any number of parameters that are available in the clear to both aggregators (e.g. information added to or already in `PAParam`).

**Comment by cjpatton** [comment]
After the design call I made two changes to #59 that might help:

1. The client now includes with its report a timestamp for report creation time. The associated data for encryption of the helper's helper's share will include this timestamp. That means the helper can authenticate the creation time for reports generated by honest clients.
2. The PA parameters now include a "batch window", which specifies the minimum time difference between the oldest and newest report in a batch. Moreover, collect requests specify the batch via a contiguous sequence of batch windows.

I wonder if this is enough plumbing to allow the helper to prevent replays without needing more overhead than a counter. Suppose for simplicity that a helper is configured for only one PA task and only handles requests from a single leader. Its encrypted state (held by the leader) could keep track of the number of times a collect request has been made for the reports in a given window. It knows the set of reports that fall in that window, so it also knows how many times each report has been used.

**Comment by csharrison** [comment]
> Its encrypted state (held by the leader) could keep track of the number of times a collect request has been made for the reports in a given window. It knows the set of reports that fall in that window, so it also knows how many times each report has been used.

Can you elaborate on exactly what the encrypted state is going to look like? It seems like your suggestion is the "stateless extreme" option I outlined in the original post. My concern is that the state will need to be a map of {batch_identifier --> counter} for all possible batch identifiers. A batch identifier could contain a time window for sure, but there may be other parameters that make this map large (collector identity, etc.).

For example, we probably want to support multiple collectors issuing queries (with different data) for batches in the same time window.

**Comment by cjpatton** [comment]
Yes, I'm advocating for the stateless extreme. My assessement is that the communication overhead will be negligible compared to the cost of re-transmitting (intermediate stages of) the IDPF shares with each aggregate request:
* IIUC the cost of the counters is `O(num_batches)` bits, where where `num_batches` is the number of batches that are valid for a collect request at any given time.  If multiple collectors are allowed to make requests for any window, then this increases to `O(num_batches * num_collectors)`. Concretely, if the batch window is 1 hour, and the leader agrees to hang on to reports for at least 24 hours, then `num_batches == 24`. (Not sure what a realistic value of `num_collectors` would be.)
* Each aggregate query includes (some intermediate stage of) the encrypted IDPF shares of a subset of reports in a batch. The size of each IDPF share is `O(n)` bits, where `n` is the number of bits of each input. That's `O(n * batch_size)` bits, where `batch_size` is the minimum number of reports per batch. Concretely, `batch_size` might be on the order of a million.

Am I missing something?

**Comment by csharrison** [comment]
Sorry for the delay on this. Yeah I think this hinges on num_collectors and num_batches. For ad-tech consumers I think this could be in the thousands+. I think num_batches may need to be larger than 24 too, if we wanted to support something like larger like weekly / monthly aggregations.

But I agree this may not be a problem in practice depending on these params and I see the benefit in avoiding premature optimization.

**Comment by cjpatton** [comment]
As discussed on today's design call, we'll resolve this issue by allowing trade-off to be made by the helper implementation. In particular, we will neither specify how the helper's state is encrypted nor "require" the helper to be stateless. (I pushed a change to #59 that reflects this resolution.)

**Comment by csharrison** [comment]
> As discussed on today's design call, we'll resolve this issue by allowing trade-off to be made by the helper implementation. In particular, we will neither specify how the helper's state is encrypted nor "require" the helper to be stateless. (I pushed a change to #59 that reflects this resolution.)

This looks good. This model supports a stateful helper and a "mostly stateless" helper that gets all its state from the cookie (It doesn't support sharding the cookie, but I think that's fine for the time being.)

Closing this out!
